applications/ShapeOptimizationApplication/python_scripts/communicator_factory.py

```diff
@@ -19,122 +19,198 @@
+    # --------------------------------------------------------------------------
+    def requestGradientOf( self, response_id ):
+        self.list_of_requests[response_id]["calculateGradient"] = True
 
-        for constraintNumber in range(numberOfConstraints):
-            constraintId = optimizationSettings["constraints"][constraintNumber]["identifier"].GetString()
-            listOfRequests[constraintId] = {"calculateValue": False, "calculateGradient": False}
+    # --------------------------------------------------------------------------
+    def isRequestingValueOf( self, response_id ):
+        if response_id not in self.list_of_requests.keys():
+            return False
+        else:
+            return self.list_of_requests[response_id]["calculateValue"]
 
-        return listOfRequests
+    # --------------------------------------------------------------------------
+    def isRequestingGradientOf( self, response_id ):
+        if response_id not in self.list_of_requests.keys():
+            return False
+        else:
+            return self.list_of_requests[response_id]["calculateGradient"]
 
     # --------------------------------------------------------------------------
-    def __initializeListOfResponses( self, optimizationSettings ):
-        
-        listOfResponses = {}  
-        numberOfObjectives = optimizationSettings["objectives"].size()
-        numberOfConstraints = optimizationSettings["constraints"].size()
+    def reportValue( self, response_id, value ):
+        self.__storeValue( response_id, value)
+        if self.__isResponseWaitingForInitialValueAsReference( response_id ):
+            self.__setValueAsReference( response_id, value )
 
-        for objectiveNumber in range(numberOfObjectives):
-            objectiveId = optimizationSettings["objectives"][objectiveNumber]["identifier"].GetString()
-            listOfResponses[objectiveId] = {}
+        standardized_value = self.__translateValueToStandardForm( response_id, value )
+        self.__storeStandardizedValue( response_id, standardized_value )
 
-        for constraintNumber in range(numberOfConstraints):
-            constraintId = optimizationSettings["constraints"][constraintNumber]["identifier"].GetString()
-            listOfResponses[constraintId] = {}  
+    # --------------------------------------------------------------------------
+    def reportGradient( self, response_id, gradient ):
+        standardized_gradient = self.__translateGradientToStandardForm( response_id, gradient )
+        self.__storeStandardizedGradient( response_id, standardized_gradient )
 
-        for responseId in listOfResponses:
-            listOfResponses[responseId] = {"value": None, "referenceValue": None, "gradient": None} 
+    # --------------------------------------------------------------------------
+    def getValue( self, response_id ):
+        return self.list_of_responses[response_id]["value"]
```

Comment:
indeed we had this before, but for now, all we can do in such a situation is to drop a meaningful error. But this happens anyways in case we ask for something that is not existing... In this context, I anyways though it might make sense to include a ValidateResponse() function which after each call of the analyzer checks all the responses against all requests and drops an error if they do not match.

Reply:
Thats also fine for me.
We just need to be careful with the ```None``` values that we have as default. Because if we pass them to somewhere else, it might be difficult to track the error

Reply:
This is what i meant